packages/peregrine/src/hooks/usePagination.js

```diff
@@ -1,14 +1,92 @@
-import { useMemo, useState } from 'react';
+import { useCallback, useMemo, useState } from 'react';
+import { useSearchParam } from './useSearchParam';
 
-export const usePagination = () => {
-    const [currentPage, setCurrentPage] = useState(0);
+/**
+ * Sets a query parameter in history. Attempt to use React Router if provided
+ * otherwise fallback to builtins.
+ */
+const setQueryParam = ({ location, history, parameter, value }) => {
+    const { search } = location;
+    const queryParams = new URLSearchParams(search);
+    queryParams.set(parameter, value);
+
+    const fn = history.push || history.pushState;
+    fn({ search: queryParams.toString() });
+};
+
+/**
+ * `usePagination` provides a pagination state with `currentPage` and
+ * `totalPages` as well as an API for interacting with the state.
+ *
+ * @param {Object} location the location object, like window.location, or from react router
+ * @param {Object} history the history object, like window.history, or from react router
+ * @param {String} namespace the namespace to apply to the pagination query
+ * @param {String} parameter the name of the query parameter to use for page
+ * @param {Number} initialPage the initial current page value
+ *
+ * TODO update with defaults
+ *
+ * @returns {[PaginationState, PaginationApi]}
+ */
+export const usePagination = ({
+    location = window.location,
+    history = window.history,
+    namespace = '',
+    parameter = 'page',
+    initialPage = 1
```

Comment:
If we're going to do a re-factor, I think we should go ahead and support a `totalPages` parameter as well. Right now we assume that pagination data is loaded async, but it would be nice to support components that already know the totalPages, so they don't have to init and immediately setTotalPages.

Reply:
Cool, will do.